app/src/main/java/org/schabi/newpipe/util/ListHelper.java

```diff
@@ -192,56 +202,55 @@ static int getDefaultResolutionIndex(final String defaultResolution,
-                                                       final List<VideoStream> videoOnlyStreams,
-                                                       final boolean ascendingOrder) {
-        final ArrayList<VideoStream> retList = new ArrayList<>();
-        final HashMap<String, VideoStream> hashMap = new HashMap<>();
-
-        if (videoOnlyStreams != null) {
-            for (final VideoStream stream : videoOnlyStreams) {
-                if (!showHigherResolutions
-                        && HIGH_RESOLUTION_LIST.contains(stream.getResolution())) {
-                    continue;
-                }
-                retList.add(stream);
-            }
-        }
-        if (videoStreams != null) {
-            for (final VideoStream stream : videoStreams) {
-                if (!showHigherResolutions
-                        && HIGH_RESOLUTION_LIST.contains(stream.getResolution())) {
-                    continue;
-                }
-                retList.add(stream);
-            }
-        }
+    @NonNull
+    static List<VideoStream> getSortedStreamVideosList(
+            @Nullable final MediaFormat defaultFormat,
+            final boolean showHigherResolutions,
+            @Nullable final List<VideoStream> videoStreams,
+            @Nullable final List<VideoStream> videoOnlyStreams,
+            final boolean ascendingOrder,
+            final boolean preferVideoOnlyStreams
+    ) {
+        // Determine order of streams
+        // The last added list is preferred
+        final List<List<VideoStream>> videoStreamsOrdered =
+                preferVideoOnlyStreams
+                        ? Arrays.asList(videoStreams, videoOnlyStreams)
+                        : Arrays.asList(videoOnlyStreams, videoStreams);
+
+        final List<VideoStream> allInitialStreams = videoStreamsOrdered.stream()
+                // Ignore lists that are null
+                .filter(Objects::nonNull)
+                .flatMap(List::stream)
+                // Filter out higher resolutions (or not if high resolutions should always be shown)
+                .filter(stream -> showHigherResolutions
+                        || !HIGH_RESOLUTION_LIST.contains(stream.getResolution()))
+                .collect(Collectors.toList());
 
+        final HashMap<String, VideoStream> hashMap = new HashMap<>();
         // Add all to the hashmap
-        for (final VideoStream videoStream : retList) {
+        for (final VideoStream videoStream : allInitialStreams) {
             hashMap.put(videoStream.getResolution(), videoStream);
         }
 
         // Override the values when the key == resolution, with the defaultFormat
-        for (final VideoStream videoStream : retList) {
+        for (final VideoStream videoStream : allInitialStreams) {
             if (videoStream.getFormat() == defaultFormat) {
                 hashMap.put(videoStream.getResolution(), videoStream);
             }
         }
 
```

Comment:
What's the purpose of this `for`? Doesn't it put in the map again some of the same items from the for above? Also, can all of this be done with streams maybe?

Reply:
I didn't change this, @litetex made these changes.

Reply:
> What's the purpose of this for?

I assume this is there to ensure that the default format is always there (by overriding potential non default ones in the hashmap).

Think of the following use case:

default format=mp4
res, format
480p, webm
480p, mp4
720p, webm
720p, mp4
1080p, webm

This should result in a list that looks like this:
480p, mp4
720p, mp4
1080p, webm


Maybe this could be improved by using java streams grouping, however I didn't touch it as it doesn't belong to this PR.

Reply:
Ok, but if you take a look at the code you see that the second `for` always adds some items that were already added by the first one. This is because the first adds them all, the second just adds a subset of the ones added by the first one since there is an additional `if`.

> Maybe this could be improved by using java streams grouping, however I didn't touch it as it doesn't belong to this PR.

Well, didn't you convert to streams the code just above?

Reply:
> Ok, but if you take a look at the code you see that the second for always adds some items that were already added by the first one. This is because the first adds them all, the second just adds a subset of the ones added by the first one since there is an additional if.

Note that the above code works with a ``HashMap``. The same keys (or in this case resolutions) will never be there twice. See also https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#put-K-V-

Okay so I let's go through this:
Assuming that default format=mp4 and ``allInitialStreams`` contains:
res, format
480p, mp4
480p, webm
720p, mp4
720p, webm
1080p, webm

After the first ``for`` the HashMap looks like this:
480p, webm
720p, webm
1080p, webm

However the user wanted default format=mp4 and that's why there is a second ``for`` which overrides the webm values with mp4. After this the map looks like this:
480p, mp4
720p, mp4
1080p, webm

> Well, didn't you convert to streams the code just above?

Yes because I couldn't build the code (thx to checkstyle-suppression file).

Reply:
> However the user wanted default format=mp4 and that's why there is a second for which overrides the webm values with mp4. After this the map looks like this:

You are right, sorry for the confusion ;-)

> Yes because I couldn't build the code (thx to checkstyle-suppression file).

I see, then these other lines shouldn't be touched in this PR